Add clearer error message using `&str + &str`

mgattozzi opened the pull request:
This is the first part of #39018. One of the common things for new users
coming from more dynamic languages like JavaScript, Python or Ruby is to
use `+` to concatenate strings. However, this doesn't work that way in
Rust unless the first type is a `String`. This commit adds a check for
this use case and outputs a new error as well as a suggestion to guide
the user towards the desired behavior. It also adds a new test case to
test the output of the error.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @nrc (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

Manishearth commented on src/librustc_typeck/check/op.rs:
```diff
@@ -238,6 +244,36 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                        };
+                    err.span_suggestion(expr.span,
+                        &format!("Your first argument should be a `String`, not `&str`. \
+                                    Try using `to_string()`"),
```
`.to_string()`

Manishearth commented on src/test/compile-fail/E0369.rs:
```diff
@@ -0,0 +1,17 @@
+    //~^ ERROR 12:18: 12:26: binary operation `+` cannot be applied to type `&'static str` [E0369]
+    //~| NOTE You can't use `+` between a `&str` and `&str`
+    //~| HELP Your first argument should be a `String`, not `&str`. Try using `to_string()`
+    //~| NOTE an implementation of `std::ops::Add` might be missing for `&'static str`
```
Have a thingy for the suggestion too

Manishearth commented:
Mostly looks okay? Would like to see if @jonathandturner has thoughts on this.

Feel free to try and add handling for the other forms. `String` is a struct so you can search for it.

I've been considering having lang-item-like annotations for types in the stdlib, which let diagnostics check for their existence. We do path matching extensively in clippy, and it would be useful to do something less hacky in rustc.

bjorn3 commented on src/librustc_typeck/check/op.rs:
```diff
@@ -221,8 +223,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                            }
                             span_note!(&mut err, lhs_expr.span,
-                                       "an implementation of `{}` might be missing for `{}`",
+                                        "an implementation of `{}` might be missing for `{}`",
```
Please remove this extra space.

mgattozzi commented:
I've updated the code with the above suggestions. @Manishearth how would I go about searching for a `String` so that I could match against it here.

Manishearth commented:
This is what we do in clippy to match structs.

https://github.com/Manishearth/rust-clippy/blob/630e0ef96875e0958277674f81921bfaba30bc5f/clippy_lints/src/utils/mod.rs#L140-L179

I think we should just land as-is since that's a bit hacky. I'm filing an issue for the diagnostic-lang-item stuff.

r? @jonathandturner

Manishearth commented:
Filed https://github.com/rust-lang/rust/issues/39131 about the diagnostic item

mgattozzi commented:
In the meantime I think it might be good to get the diagnostic of things like `String + String` if possible but that could be a separate PR if that's better.

Manishearth commented:
Actually, you can make this work without the full framework of diagnostics items; just add a `#[diagnostic_item = "String"]` annotation and then in your code check that `cx.tcx.get_attrs(adt_def.def_id)` has that attribute.

mgattozzi commented:
@Manishearth using the code from clippy or just putting it in the code now as is?

Manishearth commented:
No, we don't want to use the clippy code, it's hacky. Use an attribute as shown above.

Manishearth commented:
Maybe `rustc_diagnostic_item`

mgattozzi commented:
I used ripgrep and git grep to search the code and neither of those attributes exist in the code base so I have no basis of comparison to do it the way you've mentioned.

Manishearth commented:
I'm saying you invent a new attribute :)

Annotate String with `#[rustc_diagnostic_item = "String"]`. Then, using `tcx.get_attrs(adt_def.def_id)`, you can check if your `TyAdt` has that attribute.

You will probably have to add it to libsyntax/feature_gate.rs as well, as a Gated attribute.

Manishearth commented:
Anyway we can land as-is if you'd prefer too.

jonathandturner commented:
I like the idea.

Can you use a src/test/ui test instead of a compile_fail test?  Just so that we can see what the final output looks like a bit better in the PR.

mgattozzi commented:
@Manishearth ohhhhhh okay I think I see what you're talking about now. I can try when I'm off work.

@jonathandturner I'll switch it over to ui test when I'm off work!

nrc commented on src/librustc_typeck/check/op.rs:
```diff
@@ -238,6 +244,36 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+            if let TyRef(_, r_ty) = rhs_ty.sty {
+                if l_ty.ty.sty == TyStr && r_ty.ty.sty == TyStr {
+                    span_note!(&mut err, lhs_expr.span,
+                            "You can't use `+` between a `&str` and `&str`");
```
We usually use the passive voice for error messages, i.e., something like "`+` can't be used to concatenate two `&str` strings" and similarly later on.

mgattozzi replied:
Consistency is good. This one is a quick fix.

nrc commented on src/librustc_typeck/check/op.rs:
```diff
@@ -238,6 +244,36 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                            _ => format!("<expression>")
+                        };
+                    err.span_suggestion(expr.span,
+                        &format!("Your first argument should be a `String`, not `&str`. \
```
It would be nice to explain *why* a String is needed, otherwise this seems a little like boilerplate.

mgattozzi replied:
Yeah I can see why it's boilerplate like. I'll put a more descriptive message.

nrc commented on src/librustc_typeck/check/op.rs:
```diff
@@ -238,6 +244,36 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                        };
+                    err.span_suggestion(expr.span,
+                        &format!("Your first argument should be a `String`, not `&str`. \
+                                    Try using to_string()"),
```
I would prefer phrasing like "`to_string` can be used to create an owned string from a string reference", I actually prefer using `to_owned` rather than `to_string` because it better describes what is going on

mgattozzi replied:
`to_owned` or `to_string` is always a hot button topic. Now they're semantically equivalent at least. I can put `to_owned`. It's all the same to me and with the rest of the sentence you put it communicates the intent of the function call.

nrc commented on src/test/compile-fail/E0369.rs:
```diff
@@ -0,0 +1,18 @@
+    //~| NOTE You can't use `+` between a `&str` and `&str`
+    //~| HELP Your first argument should be a `String`, not `&str`. Try using to_string()
+    //~| SUGGESTION let string = "Hello ".to_string() + "World" + "!"
+    //~| NOTE an implementation of `std::ops::Add` might be missing for `&'static str`
```
Would be nice if we could skip the NOTE in this case.

jonathandturner replied:
@nrc - I think he's going to move to a UI test, so we won't need them (and we'll be able to see what the actual error will look like)

nrc replied:
I meant remove it from the error message, not from the test

mgattozzi replied:
Shouldn't be too hard. I have an idea of how to bypass this output.

mgattozzi commented:
@nrc I've updated the wording, though I'm wondering if it could use some improvement. I didn't want to make it too verbose, but I wanted to get the point across as well. I've also added a way to bypass the NOTE you didn't want output here through some conditional logic short circuiting. It feels a little hackish and I was wondering if you had a better idea here as to what could be improved.

Also not since the test hasn't been updated yet this should fail testing.

bjorn3 commented on src/librustc_typeck/check/op.rs:
```diff
@@ -223,13 +223,18 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                            // and it was an addition with &str then skip the
+                            // span note below it. Otherwise just emit the span
+                            // note
+                            if missing_trait == "std::ops::Add" &&
```
This will also skip the note when the trait is not Add instead of only when Add + &str

mgattozzi replied:
You're right. I got the logic wrong here. I tested it just to make sure so this will need to get changed

nrc commented on src/librustc_typeck/check/op.rs:
```diff
@@ -238,6 +248,45 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                    err.span_suggestion(expr.span,
+                        &format!("to_owned() can be used to create an owned `String` \
+                                  from a string reference. This allows concatenation \
+                                  since the `String` is owned."), suggestion);
```
The second sentence doesn't quite explain why that is necessary. How about, "String concatenation appends the string on the right to the string on the left. That modifies the string on the left and may require reallocation, this requires ownership of the string on the left." It's a bit wordy, but tries to explain why you can't use a reference.

mgattozzi commented:
@nrc and @jonathandturner these commits should cover both the ui test and message output that was brought up earlier.

mgattozzi commented:
Looking at the raw log one failed on my test due to the output being different but it passed for all the other ones. I'm not sure why this would happen though

nrc commented:
r = me with the test fixed. I'm not sure why that is failing, but it looks legit to me.

mgattozzi commented:
@nrc looking at the logs further this seems to be the problem

```
expected stderr:
error[E0514]: found crate `std` compiled by an incompatible version of rustc
  |
  = help: please recompile that crate using this compiler (rustc 1.16.0-dev (b27d71cb1 2017-01-23))
  = note: crate `std` path #1: /home/michael/Code/rustc/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-86f66de14176b4c2.rlib compiled by "rustc 1.16.0-dev (5c73f42db 2017-01-23)"
  = note: crate `std` path #2: /home/michael/Code/rustc/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-86f66de14176b4c2.so compiled by "rustc 1.16.0-dev (5c73f42db 2017-01-23)"
  = note: crate `std` path #3: /home/michael/Code/rustc/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_unicode-b421c85d3facafd6.rlib compiled by "rustc 1.16.0-dev (5c73f42db 2017-01-23)"
  = note: crate `std` path #4: /home/michael/Code/rustc/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_shim-dac73ac8aa295bd7.rlib compiled by "rustc 1.16.0-dev (5c73f42db 2017-01-23)"

error: aborting due to previous error
```

I had this happen once or twice during my builds to test all this, though I'm not sure what caused it. Usually just a clean and rebuild fixed it. Possibly restarting the build might get rid of the error.

mgattozzi commented:
@nrc I fixed the build!

nrc commented:
@mgattozzi nice! Could you squash the commits please? Then r+

mgattozzi commented:
@nrc squash and force push?

nrc commented:
yes

mgattozzi commented:
@nrc got it. Wasn't sure policy regarding force pushing. Alright it's all rebased and good to go!

nrc commented:
@bors: r+

Thanks for sticking with it @mgattozzi , nice PR!

bors commented:
:pushpin: Commit 4da2994 has been approved by `nrc`

<!-- @bors r=nrc 4da2994d86bcf8183f8d241cabd29fd6c411e628 -->

mgattozzi commented:
@nrc I'm glad to have been able to contribute to the compiler itself finally! It was a joy and I'm hoping to tackle some more issues. Thanks for reviewing it!

bors commented:
:hourglass: Testing commit 4da2994 with merge 27a80f7...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/195057921)

mgattozzi commented:
@nrc looks like curl messed up trying to download the rustc binary to do compilation.

GuillaumeGomez commented:
@bors: retry

bors commented:
:umbrella: The latest upstream changes (presumably #39353) made this pull request unmergeable. Please resolve the merge conflicts.

kevinmehall commented:
Looks like this unintentionally added files under `src/target/debug/` to git.

mgattozzi commented:
@kevinmehall thanks I'll remove it

mgattozzi commented:
@nrc it needs to get approved again so it can go through

nrc commented:
@bors: r+

bors commented:
:pushpin: Commit b54f593 has been approved by `nrc`

<!-- @bors r=nrc b54f593cffc40e9d07650b36629e60c48da6b11d -->

bors commented:
:hourglass: Testing commit b54f593 with merge 6abe648...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1784), [status-travis](https://travis-ci.org/rust-lang/rust/builds/197561140)
Approved by: nrc
Pushing 6abe64871e6e9c034d2541ee731fbe7b45893f7a to master...

Manishearth commented:
Finally

mgattozzi commented:
@Manishearth and @nrc thanks a lot! This great that it's finally merged!
